docker/docker.ts

```diff
@@ -175,36 +180,76 @@ async function buildAndPushImageWorkerAsync(
-        cacheFrom = pullCacheAsync(imageName, cacheFromParam, login, repositoryUrl, logResource);
-    } else {
-        cacheFrom = Promise.resolve(undefined);
+        cacheFrom = await pullCacheAsync(imageName, cacheFromParam, login, repositoryUrl, logResource);
     }
 
     // First build the image.
-    const buildResult = await buildImageAsync(imageName, pathOrBuild, logResource, cacheFrom);
+    const { imageId, stages } = await buildImageAsync(imageName, pathOrBuild, logResource, cacheFrom);
+
+    if (!repositoryUrl) {
+        throw new ResourceError("Expected repository URL to be defined during push", logResource);
+    }
 
-    // Use those then push the image.  Then just return the digest as the final result for our caller to use.
+    // Generate a name that uniquely will identify this built image.  This is similar in purpose to
+    // the name@digest form that can be normally be retrieved from a docker repository.  However,
+    // this tag doesn't require actually pushing the image, nor does it require communicating with
+    // some external system, making it suitable for unique identification, even during preview.
+    // This also means that if docker produces a new imageId, we'll get a new name here, ensuring that
+    // resources (like docker.Image and cloud.Service) will be appropriately replaced.
+    const uniqueTargetName = createTargetName(repositoryUrl, tag, imageId);
+
+    // Use those then push the image.  Then just return the unique target name. as the final result
+    // for our caller to use.
     if (!pulumi.runtime.isDryRun()) {
-        // Only push the image during an update, do not push during a preview, even if digest and url are available
-        // from a previous update.
+        // Only push the image during an update, do not push during a preview.
         await login();
 
         // Push the final image first, then push the stage images to use for caching.
-        await pushImageAsync(imageName, repositoryUrl, logResource);
+        await tagAndPushImageAsync(imageName, repositoryUrl, tag, imageId, logResource);
 
-        for (const stage of buildResult.stages) {
-            await pushImageAsync(
-                localStageImageName(imageName, stage), repositoryUrl, logResource, stage);
+        for (const stage of stages) {
+            await tagAndPushImageAsync(
+                localStageImageName(imageName, stage), repositoryUrl, tag, /*imageId:*/ undefined, logResource);
         }
     }
 
-    return buildResult.id;
+    return uniqueTargetName;
+}
+
+function localStageImageName(imageName: string, stage: string) {
+    return `${imageName}-${stage}`;
+}
+
+function getImageNameAndTag(baseImageName: string): { imageName: string, tag: string | undefined } {
+    const lastColon = baseImageName.lastIndexOf(":");
+    const imageName = lastColon < 0 ? baseImageName : baseImageName.substr(0, lastColon);
+    const tag = lastColon < 0 ? undefined : baseImageName.substr(lastColon + 1);
```

Comment:
Nit: I think `.substr` is non-standard - and that `.substring` is the preferred function. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr

But super-confusingly - they are different - `substring` second param is end not length - though that shouldn't matter for the two uses above.

Reply:
i actually love JS for this.  i think it's good they have two functions, since sometimes start/length is convenient, and sometimes start/end is convenient.  :)